### DISCUSSION TITLE
Drain mode when reducing numprocs

### DISCUSSION THREAD

**Discussion by BrianTAllen** [julien6387/supvisors#116] [General]
Hi,

Thanks for the great plugin!

When reducing numprocs it kills running processes to get down to the new numprocs amount.

Is there a way to do a drain mode so when you reduce the numprocs below the number of running processes the existing process will continue to run until they complete and exit on their own, and then new processes won't respawn (until the number of running processes eventually goes below numprocs, and then they'll respawn accordingly to get back up to numprocs)?

Thanks!

**Comment by julien6387** [comment]
Hi.

Well, at first sight there should be a way.
The process could be marked for removal if running and impacted by a numprocs decrease.
Upon reception of a stopped-like internal notification from Supervisor, such a marked process could be removed from Supervisor before Supervisor triggers any autorestart behaviour (through a subsequent call to its transition method).
The restart on the Web UI would have to be disabled too in this case.

Increasing the numprocs is a bit more complicated in such a case.
Nothing infeasible anyway.

I'm more worried about the impact on other Supvisors features.
There would be a mid-term discrepancy between the process data and the program options.
Not 100% sure that everything is robust to 'holes' in homogeneous process groups.

I may give it a try someday.

**Reply by BrianTAllen** [reply]
Thanks for the insights and reply.

**Comment by julien6387** [comment]
Hi,

I have re-designed the function and added a new XML-RPC `lazy_update_numprocs` in the branch dev-0.18.
(I must admit that my mind was lazy too when I was looking for a name ^^)
Still some testing to do but it's stable enough to share.

Cheers,.

**Comment by julien6387** [comment]
implemented in Supvisors 0.18 with a `lazy` attribute in the existing `update_numprocs` XML-RPC

**Reply by BrianTAllen** [reply]
Thank you! I'll check it out
